Empty file in the spool dir prevents extsmaild from sending mails

Hi Laurie,

I just ran into this issue: I was sending mails to my address but did not receive them.
A `ls -l ~/.extsmail/spool_dir/msgs/` showed an empty file and my mails.
Removing the empty file and restarting `extsmaild` solved the issue.

I did a quick test, it's better (i now get 3 of 5 mails instead of 2):

```
$ cd ~/.extsmail/spool_dir/msgs
$ for i in `seq 1 5`; do echo `date` | mail -s "test extsmail $i" olivier@biniou.info; done
# this is OK
$ touch foobar
$ for i in `seq 1 5`; do echo `date` | mail -s "test extsmail $i" olivier@biniou.info; done
# i get 3 mails
$ ls -l
total 8.0K
-rw-r--r-- 1 olivier olivier   0 Jun 27 21:05 foobar
-rw------- 1 olivier olivier 126 Jun 27 21:05 XXXXhAEyIF
-rw------- 1 olivier olivier 126 Jun 27 21:05 XXXXxAvXYK
$ for i in `seq 1 5`; do echo `date` | mail -s "test extsmail $i" olivier@biniou.info; done
# i get 3 mails
$ ls -l
total 16K
-rw-r--r-- 1 olivier olivier   0 Jun 27 21:05 foobar
-rw------- 1 olivier olivier 126 Jun 27 21:05 XXXXNgVIZp
-rw------- 1 olivier olivier 126 Jun 27 21:05 XXXXrsBDrH
-rw------- 1 olivier olivier 126 Jun 27 21:05 XXXXsL9dpN
-rw------- 1 olivier olivier 126 Jun 27 21:05 XXXXxYysc7
```

BTW there's a gcc warning:
```
extsmaild.c: In function ‘cycle’:
extsmaild.c:587:1: warning: label ‘next_msg’ defined but not used [-Wunused-label]
  587 | next_msg:
      | ^~~~~~~~
```

--- a/configure.ac
+++ b/configure.ac
@@ -72,6 +72,11 @@ esac
 
 LDFLAGS="$LDFLAGS -Wl,-z,relro,-z,now"
 
+# arc4random_uniform
+
+AH_TEMPLATE(HAVE_ARC4RANDOM_UNIFORM, [Define if your platform has the arc4random_uniform function.])
+AC_CHECK_FUNC(arc4random_uniform, [AC_DEFINE(HAVE_ARC4RANDOM_UNIFORM)])
+
 # strtonum
 
 AH_TEMPLATE(HAVE_STRTONUM, [Define if your platform has the strtonum function.])

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -93,7 +93,6 @@ typedef struct _pid_llist {
 } PID_LList;
 
 typedef struct {
-    long spool_loc;       // The next entry number in the spool dir to try.
     bool any_failure;     // Set to true if any message in the spool dir was not
                           // sent. Reset on each cycle
     time_t last_success;  // The time of the last successful send of all
@@ -397,45 +396,24 @@ static int cycle(Conf *conf, Group *groups, Status *status)
     }
 
     if (conf->mode == DAEMON_MODE) {
-        // In daemon mode, we try to make sure we don't get "stuck" on the same
-        // message. For example, if a message is too big to be sent by the
-        // remote sendmail, it could stop us trying to send other messages which
-        // might be successful. So, on each cycle, we try starting one entry
-        // further into the spool dir. In other words if we have entries:
-        //   a, b, c
-        // in the spool dir then we will try sending them in this order upon
-        // each cycle (until some are actually sent):
-        //   1: a, b, c
-        //   2: b, c, a
-        //   3: c, a, b
-        // Note that this is only a "best effort": the order that files are
-        // returned by readdir() is undefined, though on most operating systems
-        // it is likely to be stable until files are added or removed.
-
-        // Skip the entries we tried on the last iteration.
-        for (int i = 0; i < status->spool_loc; i += 1) {
-            errno = 0;
-            if (readdir(dirp) == NULL) {
-                // We've hit the end of the directory; it probably means that the
-                // user has removed files.
-                status->spool_loc = 0;
-                if (errno == 0) {
-                    rewinddir(dirp);
-                    break;
-                }
-                else {
-                    // Something odd happened, and it's not obvious what we could
-                    // do to recover. We're best off starting afresh in the next
-                    // cycle.
-                    closedir(dirp);
-                    free(msgs_path);
-                    return false;
-                }
+        if (status->any_failure) {
+            // In daemon mode, we try to make sure we don't get "stuck" on the
+            // same message. For example, if a message is too big to be sent by
+            // the remote sendmail, it can stop us sending other messages. If
+            // in the last cycle we encountered a failure, we randomly skip
+            // some entries in the queue directory on the basis that it might
+            // allow some other messages to be sent.
+            uint32_t num_entries = 0;
+            while (readdir(dirp) != NULL) {
+                num_entries += 1;
+            }
+            rewinddir(dirp);
+            uint32_t skip = arc4random_uniform(num_entries);
+            for (uint32_t i = 0; i < skip; i += 1) {
+                readdir(dirp);
             }
         }
     }
-    long start_spool_loc;
-    start_spool_loc = status->spool_loc;
 
     // Reset all the externals "working" status so that we'll try all of them
     // again.
@@ -451,7 +429,7 @@ static int cycle(Conf *conf, Group *groups, Status *status)
     }
 
     bool all_sent = true;
-    bool tried_once = false; // Make sure we've tried at least one file.
+    bool reached_end = false; // Have we iterated through the directory at least once?
     int num_successes = 0;   // How many messages have been successfully sent.
     while (1) {
         char *msg_path = NULL;
@@ -462,32 +440,11 @@ static int cycle(Conf *conf, Group *groups, Status *status)
             if (errno == 0) {
                 // We've got to the end of the directory.
                 if (conf->mode == DAEMON_MODE) {
-                    if (tried_once && start_spool_loc == 0) {
-                        // We started this cycle from the start of the directory,
-                        // so we can now be sure that we've read everything.
+                    if (reached_end
+                      || (!reached_end && !status->any_failure && num_successes == 0)) {
                         break;
                     }
-                    else if (!tried_once) {
-                        // We haven't read any entries, but we've already read
-                        // past the end of the directory. We reset the directory
-                        // read to the start and carry on as if we'd always
-                        // intended to start from the beginning of the
-                        // directory.
-                        start_spool_loc = 0;
-                    }
-                    else if (num_successes > 0
-                          || start_spool_loc > status->spool_loc) {
-                        // Entries have been removed from the directory during
-                        // the cycle (probably because we've sent messages
-                        // successfully, but maybe because of user interaction).
-                        // Either way, this is likely to change the order of
-                        // files returned by readdir(), so it's better to try
-                        // everything in the directory again.
-                        start_spool_loc = 0;
-                        tried_once = false;
-                    }
-
-                    status->spool_loc = 0;
+                    reached_end = true;
                     rewinddir(dirp);
                     continue;
                 }
@@ -502,7 +459,6 @@ static int cycle(Conf *conf, Group *groups, Status *status)
 
         // The entries "." and ".." are, fairly obviously, not messages.
         if (strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0) {
-            status->spool_loc += 1;
             continue;
         }
 
@@ -576,27 +532,6 @@ next_try:
         }
 
         free(msg_path);
-        if (conf->mode == DAEMON_MODE) {
-            if (!all_sent
-              || (tried_once && start_spool_loc == status->spool_loc)) {
-                // Either a message send has failed or we've got back to the
-                // point in the spool dir where we started.
-                if (num_successes > 0) {
-                    // If one or more messages was sent, then the removal of those
-                    // files from the file system is likely to perturb the order of
-                    // files in readdir(), such that spool_loc is likely to be
-                    // misleading. It's better to start from the beginning of the
-                    // directory next time.
-                    status->spool_loc = 0;
-                } else {
-                    status->spool_loc += 1;
-                }
-                break;
-            }
-
-            status->spool_loc += 1;
-            tried_once = true;
-        }
     }
 
     closedir(dirp);
@@ -604,7 +539,6 @@ next_try:
 
     if (conf->mode == DAEMON_MODE) {
         if (all_sent) {
-            status->spool_loc = 0;
             status->any_failure = false;
             status->last_success = status->last_notify_failure = time(NULL);
             if (num_successes == 0) {
@@ -1567,7 +1501,6 @@ int main(int argc, char** argv)
         exit(1);
 
     Status status;
-    status.spool_loc = 0;
     status.any_failure = false;
     status.last_success = status.last_notify_failure = time(NULL);
     status.pid_llist = NULL;

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -449,7 +449,7 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
     }
 
     bool all_sent = true;
-    bool tried_once = false; // Make sure we read every entry at least once.
+    bool tried_once = false; // Make sure we've tried at least one file.
     int num_successes = 0;   // How many messages have been successfully sent.
     while (1) {
         char *msg_path = NULL;

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -472,69 +472,54 @@ static int cycle(Conf *conf, Group *groups, Status *status)
             exit(1);
         }
 
-        int tries = 3; // Max number of times we'll try to operate on this file.
-        while (1) {
-            int fd = open(msg_path, O_RDONLY);
-            if (fd == -1) {
-                // If we couldn't so much as open the file then something odd
-                // has happened.
-                goto next_try;
-            }
-
-            if (flock(fd, LOCK_EX | LOCK_NB) == -1) {
-                if (errno != EWOULDBLOCK)
-                    syslog(LOG_ERR, "Error when flocking'ing '%s'", msg_path);
-
-                close(fd);
-                goto next_try;
-            }
-
-            // We've now got an exclusive lock on the file.
-            struct stat msg_st;
-            if (fstat(fd, &msg_st) == -1) {
-                syslog(LOG_ERR, "Error when fstat'ing '%s'", msg_path);
-                close(fd);
-                goto next_try;
-            }
-
-            // If the file is zero size it probably means that we managed to
-            // interrupt extsmail before it had managed to obtain an
-            // exclusive lock and write any data to the file. In such a case
-            // we don't try and do anything; instead (below) we sleep a
-            // second and then try again. This is because in general we
-            // assume that extsmail will have finished with a file in at
-            // most a couple of seconds (and probably much less).
-
-            if (msg_st.st_size == 0) {
-                close(fd);
-                goto next_try;
-            }
-
-            if (try_groups(conf, status, msg_path, fd)) {
-                num_successes += 1;
-                break;
-            }
-
-            // The send failed for whatever reason. Give up for the time being.
+        int fd = open(msg_path, O_RDONLY);
+        if (fd == -1) {
+            // If we couldn't so much as open the file then something odd
+            // has happened.
+            free(msg_path);
             all_sent = false;
-            close(fd);
-            break;
-
-next_try:
-            // At this point, either we've released the exclusive lock we'd
-            // previously gained (because we'd gained it before any data had
-            // been written to the spool file) or we weren't able to gain the
-            // lock at all. Assuming we haven't tried this too many times, we now
-            // try sleeping for a second and then having another go.
-
-            tries -= 1;
-            if (tries < 0) {
-                all_sent = false;
-                break;
-            }
-            sleep(1);
+            continue;
         }
 
+        if (flock(fd, LOCK_EX | LOCK_NB) == -1) {
+            if (errno != EWOULDBLOCK)
+                syslog(LOG_ERR, "Error when flocking'ing '%s'", msg_path);
+
+            free(msg_path);
+            close(fd);
+            all_sent = false;
+            continue;
+        }
+
+        // We've now got an exclusive lock on the file.
+        struct stat msg_st;
+        if (fstat(fd, &msg_st) == -1) {
+            syslog(LOG_ERR, "Error when fstat'ing '%s'", msg_path);
+            free(msg_path);
+            close(fd);
+            all_sent = false;
+            continue;
+        }
+
+       if (msg_st.st_size == 0) {
+            // If the file is zero size it either means that we managed to
+            // interrupt extsmail before it had managed to obtain an exclusive
+            // lock and write any data to the file or the user accidentally
+            // created a 0 length file in the spool dir (which is easy to do by
+            // calling "extsmail" instead of "extsmaild"). Because the latter
+            // is a semi-permanent condition, we don't count this as
+            // unsuccessful.
+            free(msg_path);
+            close(fd);
+            continue;
+        }
+
+        if (try_groups(conf, status, msg_path, fd)) {
+            num_successes += 1;
+        } else {
+            all_sent = false;
+            close(fd);
+        }
         free(msg_path);
     }
 

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -577,14 +577,11 @@ next_try:
             sleep(1);
         }
 
+        free(msg_path);
         if (conf->mode == DAEMON_MODE && !all_sent) {
             status->spool_loc = spool_loc + 1;
-            free(msg_path);
             break;
         }
-
-        if (msg_path)
-            free(msg_path);
 
         if (conf->mode == DAEMON_MODE) {
             if (tried_once && spool_loc == start_spool_loc) {

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -408,7 +408,11 @@ static int cycle(Conf *conf, Group *groups, Status *status)
                 num_entries += 1;
             }
             rewinddir(dirp);
+#ifdef HAVE_ARC4RANDOM_UNIFORM
             uint32_t skip = arc4random_uniform(num_entries);
+#else
+            uint32_t skip = ((uint32_t) rand()) % num_entries;
+#endif
             for (uint32_t i = 0; i < skip; i += 1) {
                 readdir(dirp);
             }

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -409,13 +409,16 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
         //   1: a, b, c
         //   2: b, c, a
         //   3: c, a, b
+        // Note that this is only a "best effort": the order that files are
+        // returned by readdir() is undefined, though on most operating systems
+        // it is likely to be stable until files are added or removed.
 
         // Skip the entries we tried on the last iteration.
         for (int i = 0; i < status->spool_loc; i += 1) {
             errno = 0;
             if (readdir(dirp) == NULL) {
-                // We've hit the end of the directory; it probably means that we
-                // managed to send some messages on the last iteration.
+                // We've hit the end of the directory; it probably means that the
+                // user has removed files.
                 status->spool_loc = 0;
                 if (errno == 0) {
                     rewinddir(dirp);
@@ -471,24 +474,21 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
                         // read to the start and carry on as if we'd always
                         // intended to start from the beginning of the
                         // directory.
-                        start_spool_loc = status->spool_loc = 0;
+                        start_spool_loc = 0;
                     }
-                    else if (start_spool_loc > status->spool_loc) {
+                    else if (num_successes > 0
+                          || start_spool_loc > status->spool_loc) {
                         // Entries have been removed from the directory during
                         // the cycle (probably because we've sent messages
                         // successfully, but maybe because of user interaction).
-                        // We reset the directory read to the start, in case
-                        // files were present earlier in the readdir that we
-                        // never tried, or new files have been added in the
-                        // interim.
-                        start_spool_loc = status->spool_loc = 0;
+                        // Either way, this is likely to change the order of
+                        // files returned by readdir(), so it's better to try
+                        // everything in the directory again.
+                        start_spool_loc = 0;
                         tried_once = false;
                     }
 
-                    // There could be entries between seekdir(0) and
-                    // seekdir(status->spool_loc) that we haven't yet tried to
-                    // send, so rewind to make sure we have a chance of trying
-                    // then.
+                    status->spool_loc = 0;
                     rewinddir(dirp);
                     continue;
                 }
@@ -577,16 +577,24 @@ next_try:
         }
 
         free(msg_path);
-        if (conf->mode == DAEMON_MODE && !all_sent) {
-            status->spool_loc += 1;
-            break;
-        }
-
         if (conf->mode == DAEMON_MODE) {
-            if (tried_once && start_spool_loc == status->spool_loc) {
-                // We've read all the directory entries at least once.
+            if (!all_sent
+              || (tried_once && start_spool_loc == status->spool_loc)) {
+                // Either a message send has failed or we've got back to the
+                // point in the spool dir where we started.
+                if (num_successes > 0) {
+                    // If one or more messages was sent, then the removal of those
+                    // files from the file system is likely to perturb the order of
+                    // files in readdir(), such that spool_loc is likely to be
+                    // misleading. It's better to start from the beginning of the
+                    // directory next time.
+                    status->spool_loc = 0;
+                } else {
+                    status->spool_loc += 1;
+                }
                 break;
             }
+
             status->spool_loc += 1;
             tried_once = true;
         }

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -432,8 +432,8 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
             }
         }
     }
-    long start_spool_loc, spool_loc;
-    start_spool_loc = spool_loc = status->spool_loc;
+    long start_spool_loc;
+    start_spool_loc = status->spool_loc;
 
     // Reset all the externals "working" status so that we'll try all of them
     // again.
@@ -473,7 +473,7 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
                         // directory.
                         start_spool_loc = status->spool_loc = 0;
                     }
-                    else if (start_spool_loc > spool_loc) {
+                    else if (start_spool_loc > status->spool_loc) {
                         // Entries have been removed from the directory during
                         // the cycle (probably because we've sent messages
                         // successfully, but maybe because of user interaction).
@@ -490,7 +490,6 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
                     // send, so rewind to make sure we have a chance of trying
                     // then.
                     rewinddir(dirp);
-                    spool_loc = 0;
                     continue;
                 }
                 else
@@ -504,7 +503,7 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
 
         // The entries "." and ".." are, fairly obviously, not messages.
         if (strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0) {
-            spool_loc += 1;
+            status->spool_loc += 1;
             continue;
         }
 
@@ -579,16 +578,16 @@ next_try:
 
         free(msg_path);
         if (conf->mode == DAEMON_MODE && !all_sent) {
-            status->spool_loc = spool_loc + 1;
+            status->spool_loc += 1;
             break;
         }
 
         if (conf->mode == DAEMON_MODE) {
-            if (tried_once && spool_loc == start_spool_loc) {
+            if (tried_once && start_spool_loc == status->spool_loc) {
                 // We've read all the directory entries at least once.
                 break;
             }
-            spool_loc += 1;
+            status->spool_loc += 1;
             tried_once = true;
         }
     }

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -116,7 +116,7 @@ extern char* __progname;
 static void read_externals(void);
 static void free_groups();
 static int try_externals_path(const char *);
-static bool cycle(Conf *, Group *, Status *);
+static int cycle(Conf *, Group *, Status *);
 static bool try_groups(Conf *, Status *, const char *, int);
 static void push_killed_pid(Status *, pid_t);
 static void cycle_killed_pids(Status *);
@@ -376,12 +376,11 @@ static int try_externals_path(const char *path)
 //
 
 //
-// Try sending any messages in the spool dir. Returns true if all messages were
-// sent successfully (or if there were no messages to send), or false if at least
-// one message failed to be succesfully sent.
+// Try sending any messages in the spool dir. Returns the number of messages
+// successfully sent.
 //
 
-static bool cycle(Conf *conf, Group *groups, Status *status)
+static int cycle(Conf *conf, Group *groups, Status *status)
 {
     char *msgs_path; // msgs dir (within spool dir)
     if (asprintf(&msgs_path, "%s%s%s", conf->spool_dir, DIR_SEP, MSGS_DIR)
@@ -632,7 +631,7 @@ next_try:
     if (num_successes > 0)
         do_notify_success_cmd(conf, num_successes);
 
-    return all_sent;
+    return num_successes;
 }
 
 
@@ -1642,9 +1641,9 @@ int main(int argc, char** argv)
 
             // The main message sending cycle
 
-            bool all_sent = cycle(conf, groups, &status);
+            int num_successes = cycle(conf, groups, &status);
 
-            if (!all_sent && conf->notify_failure_interval > 0
+            if (status.any_failure && conf->notify_failure_interval > 0
               && time(NULL) >
               status.last_notify_failure + conf->notify_failure_interval) {
                 do_notify_failure_cmd(conf, &status);
@@ -1661,10 +1660,16 @@ int main(int argc, char** argv)
             // INITIAL_POLL_WAIT to MAX_POLL_WAIT).
 
             int poll_wait;
-            if (all_sent)
-                poll_wait = MAX_POLL_WAIT;
-            else
+            if (status.any_failure) {
+                if (num_successes > 0) {
+                    // Some messages were sent, so it's better not to wait too
+                    // long before trying to send more.
+                    unsuccessful_wait = INITIAL_POLL_WAIT;
+                }
                 poll_wait = unsuccessful_wait;
+            }
+            else
+                poll_wait = MAX_POLL_WAIT;
 
             // On platforms that support an appropriate mechanism (such as kqueue
             // or inotify), we try and send messages as soon as we notice changes
@@ -1700,7 +1705,7 @@ int main(int argc, char** argv)
             sleep(poll_wait);
 #endif
 
-            if (all_sent)
+            if (!status.any_failure)
                 unsuccessful_wait = INITIAL_POLL_WAIT;
             else {
                 unsuccessful_wait *= 2;

--- a/extsmaild.c
+++ b/extsmaild.c
@@ -503,9 +503,10 @@ static bool cycle(Conf *conf, Group *groups, Status *status)
         }
 
         // The entries "." and ".." are, fairly obviously, not messages.
-
-        if (strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0)
-            goto next_msg;
+        if (strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0) {
+            spool_loc += 1;
+            continue;
+        }
 
         if (asprintf(&msg_path, "%s%s%s%s%s", conf->spool_dir, DIR_SEP,
           MSGS_DIR, DIR_SEP, dp->d_name) == -1) {
@@ -582,7 +583,6 @@ next_try:
             break;
         }
 
-next_msg:
         if (msg_path)
             free(msg_path);
 

